lib/puppet/provider/vcsrepo/git.rb

```diff
@@ -423,16 +423,17 @@ def commits?
   # handle upstream branch changes
   # @!visibility private
   def checkout(revision = @resource.value(:revision))
+    has_changes = uncommitted_changes?
     keep_local_changes = @resource.value(:keep_local_changes)
-    stash if keep_local_changes == :true
+    stash if has_changes == :true && keep_local_changes == :true
     if !local_branch_revision?(revision) && remote_branch_revision?(revision)
       # non-locally existant branches (perhaps switching to a branch that has never been checked out)
       at_path { git_with_identity('checkout', '--force', '-b', revision, '--track', "#{@resource.value(:remote)}/#{revision}") }
     else
       # tags, locally existant branches (perhaps outdated), and shas
       at_path { git_with_identity('checkout', '--force', revision) }
     end
-    unstash if keep_local_changes == :true
+    unstash if has_changes == :true && keep_local_changes == :true
```

Comment:
```suggestion
    unstash if has_changes && keep_local_changes == :true
```

Reply:
Thanks @kenyon for the quick review. I have now pushed a commit to address your suggestion.